### PR TITLE
ENT-4252: Migrate from nose to pytest

### PR DESCRIPTION
**Pull request by m-horky**
* Card ID: ENT-4252

This commit replaces nosetests with pytest as a test suite.

The DBus tests have been disabled, because they fail when not run under
nosetests, and will have to be fixed before they are enabled again. They
can be selected manually by using the '-m dbus' flag.

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -5,6 +5,24 @@ exclude=*certdata*,*manifestdata*
...
+	dbus: subscription-manager tests for DBus.
+	slow: subscription-manager tests that may be slower than the rest.
+	zypper: subscription-manager tests for the Zypper package manager.
+	functional: subscription-manager functional tests.
+timeout = 5
+testpaths =
+	test/
+required_plugins =
+	pytest-randomly
+	pytest-timeout
```
Comment (jirihnidek) [inline tox.ini:28]: To be honest, we IMHO don't use `tox` at all. Could it be useful for anything?

**Review by jirihnidek — CHANGES_REQUESTED** [review]
Thanks for the PR :+1: Overall I like it. I have few minor requests and suggestions.

**Comment by jirihnidek** [comment]
One more thing: I can see that `nosetest` is still referenced at many places. E.g. `nosetests.xml` is generated, when unit tests are run on Jenkins.

File: TESTING.md [hunk 2]
```diff
@@ -1,3 +1,78 @@
...
+```bash
+PYTHONPATH=src/ pytest -n 4 --no-summary -p no:randomly -v test/
+```
+
+After all the tests are run, a warnings summary is displayed with the list of deprecations. It can be disabled with `--disable-warnings`. Whole summary can be disabled with `--no-summary`.
+
+To compute coverage, run
+
+```bash
+coverage run -m pytest -v test/
```
[the remark this answers is not in the document]
Comment (m-horky) [reply TESTING.md:66]: It seems like the `rhsmlib` directory is not being included in `.coveragerc`. Should I add it as part of this PR?

File: tox.ini (hunk 1 above)
Comment (m-horky) [reply tox.ini:28]: We do not, but most of the tools try to read their configuration from here. IMHO it's better to keep them all in one place instead of having them all separated in their own configs (which we are currently doing for flake8 with its `.flake8`).

We can use `.pytest.ini` as a source of configuration for pytest, but that's another file in the project root. In this sense I'd vote to move the flake8 config here to the `tox.ini` in near future, to keep them all together.

File: jenkins/python3-tests.sh [hunk 3]
```diff
@@ -29,13 +24,10 @@ pip install -I -r test-requirements.txt
...
 
-# not using "setup.py nosetests" yet
-# since they need a running candlepin
+export PYTHONPATH=src/
 
-# Run just the unit tests, functional needs a running candlepin
-#pushd test/unit
 # make sure we have a dbus session for the dbus tests
-dbus-run-session nosetests --with-xunit --with-cover --cover-package rhsm --cover-package subscription_manager --cover-erase
+dbus-run-session coverage run -m pytest test/
```
[the remark this answers is not in the document]
Comment (m-horky) [reply jenkins/python3-tests.sh:30]: Well, we are planning on enabling them in a near future. Because running `dbus-run-session` means little to no difference of time/resources, I thought we may just leave it here. On the other hand, we can remove it now and put it back in once the dbus tests are corrected; we just have to remember to do that.

File: TESTING.md [hunk 4]
```diff
@@ -1,3 +1,78 @@
...
+# or, for increased verbosity
+pytest -v
+```
+
+Some tests are disabled by default, some can be disabled manually. Defaults are stored in the `tox.ini` file under `[pytest]` section. To overwrite them, use `-m` argument.
+
+For example, to run just the Zypper tests, execute
+
+```bash
+pytest -m zypper
```
[the remark this answers is not in the document]
Comment (m-horky) [reply TESTING.md:24]: Sure.

File: TESTING.md [hunk 5]
```diff
@@ -1,3 +1,78 @@
...
+```
+
+After all the tests are run, a warnings summary is displayed with the list of deprecations. It can be disabled with `--disable-warnings`. Whole summary can be disabled with `--no-summary`.
+
+To compute coverage, run
+
+```bash
+coverage run -m pytest -v test/
+# display ASCII report
+coverage report
```
[the remark this answers is not in the document]
Comment (m-horky) [reply TESTING.md:68]: Sure.

**Comment by m-horky** [comment]
I kept the `nosetests.xml` for backwards compatibility, I wasn't sure if some other system wasn't picking them up and using them somewhere.

File: TESTING.md [hunk 6]
```diff
@@ -1,3 +1,78 @@
...
+# generate interactive HTML report to htmlcov/
+coverage html
+```
+
```
[the remark this answers is not in the document]
Comment (m-horky) [reply TESTING.md:72]: I don't use PyCharm, but I'll believe you that the shortcut is correct :)

File: test/__init__.py [hunk 7]
```diff
@@ -1,5 +1,38 @@
...
+        if v not in os.environ
+    ]
+
+    skip_func: Callable = pytest.mark.skipif(
+        missing_vars,
+        reason=f"Missing Zypper environment variables {', '.join(missing_vars)}."
+    )
+    mark_func: Callable = pytest.mark.zypper
+
+    return skip_func(mark_func(func))
```
Comment (ptoscano) [inline test/__init__.py:38]: while the idea of this is great, I fear it is too specific and will be easily outdated

since the test this applies to requires environment variables, I'd rather add a `skipifnotenvvar('ENV_VAR')` function/decorator that skips a test if an environment variable is not set:
- we can add all the ones needed to the zypper test (pytests checks all the skip checks decorated to a function)
- it can possibly be used in other tests if needed
- it should be smaller and easier to maintain

having this kind of checks done as decorator would allow to drop almost all the content of `TestServicePlugin.setUp()`

File: test/rhsmlib_test/base.py [hunk 8]
```diff
@@ -108,7 +108,7 @@ def setUp(self):
         self.addCleanup(self.stop_server)
         sender_patcher = mock.patch("rhsmlib.client_info.DBusSender.get_cmd_line")
         self.mock_sender_get_cmd_line = sender_patcher.start()
-        self.mock_sender_get_cmd_line.return_value = "nose-unit-test"
+        self.mock_sender_get_cmd_line.return_value = "pytest-unit-test"
```
Comment (ptoscano) [inline test/rhsmlib_test/base.py:111]: maybe we can simply drop the name of the test runner and make it simply "unit-test", as neither nose nor pytest matter for this

File: test/test_dnf_content_plugin.py [hunk 9]
```diff
@@ -6,13 +6,16 @@
...
 
+skip: bool = False
+import_error: str = ""
 try:
     import dnf
     import librepo
 except ImportError as e:
-    raise SkipTest(e)
+    skip = True
+    import_error = str(e)
```
Comment (ptoscano) [inline test/test_dnf_content_plugin.py:18]: hm what about using `pytest.skip(msg, allow_module_level=True)` to skip the test from a file level? ie something like:
```python
try:
    import dnf
    import librepo
except ImportError as e:
    pytest.skip(f"DNF dependency could not be imported: {e}", allow_module_level=True)
```
this way, we do not need to decorate new classes added to this test module

File: test/test_dnf_content_plugin.py [hunk 10]
```diff
@@ -35,11 +38,13 @@
 try:
     dnf_product_id = imp.load_module('dnf_product_id', fp, pathname, description)
 except ImportError as e:
-    raise SkipTest(e)
+    skip = True
+    import_error = str(e)
```
Comment (ptoscano) [inline test/test_dnf_content_plugin.py:42]: ditto

File: tox.ini (hunk 1 above)
Comment (ptoscano) [reply tox.ini:28]: personally, i'd prefer to have each tool using its own configuration file, as it makes easier to see where the configuration for each lives; i already made the mistake once with `.flake8`...

File: test/zypper_test/test_serviceplugin.py [hunk 11]
```diff
@@ -6,10 +6,10 @@
 import subprocess
 import tempfile
 
-from nose.plugins.attrib import attr
+from test import subman_marker_zypper
 
 
-@attr('zypper')
+@subman_marker_zypper
 class TestServicePlugin(TestCase):
```
Comment (ptoscano) [inline test/zypper_test/test_serviceplugin.py:13]: this also needs a "functional" marker, since this is a functional test (it runs sub-man as installed, including the zypper plugin); can be an own commit

File: tox.ini [hunk 12]
```diff
@@ -5,6 +5,24 @@ exclude=*certdata*,*manifestdata*
...
+[pytest]
+addopts =
+	# Non-registered markers raise errors
+	--strict-markers
+	# Start with tests that failed last time, --ff
+	--failed-first
+	# Show extra summary for (f)ailed, (E)rror, (s)kipped and (w)arnings
+	-rfEsw
+	# Mark some tests not to be run
+	-m "not zypper and not dbus and not functional"
```
Comment (ptoscano) [inline tox.ini:17]: with my hints mentioned previously on skipping the zypper test by checking for its required environment variables (and marking it as functional too), there is no more need to exclude it by default, as it will be skipped gracefully

**Review by ptoscano — CHANGES_REQUESTED** [review]
Generally a good improvement over the current status quo :+1: There are some things which IMHO we can improve, I left notes on them.

One more general thing is about the run of the tests: the `pytest` command in setup.py tweaks `sys.path` appropriately, so why cannot the test suite itself do it on its own? for example, something like:
```python
rootdir = os.path.dirname(os.path.dirname(__file__))
sys.path.insert(0, os.path.join(rootdir, 'src'))
```
in `__init__.py`, perhaps; this way, there is no more need to export `PYTHONPATH` manually anymore (or document it), you can simply run `pytest` and that would do it :)

File: test-requirements.txt [hunk 13]
```diff
@@ -5,14 +5,13 @@
...
+pytest-randomly
+pytest-timeout
 coverage
 polib
 pyinotify
 simplejson
 mock
 Sphinx==1.8.5;python_version<="2.7"
 Sphinx>=1.8.5;python_version>="3.0"
-git+https://github.com/awood/nose-xvfb.git#egg=nose-xvfb
```
[the remark this answers is not in the document]
Comment (m-horky) [reply test-requirements.txt:18]: It is true that it is used to test GUI, but it is a nosetest plugin, I believe this is a right time for this to be removed.

File: TESTING.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (m-horky) [reply TESTING.md:24]: After @ptoscano's suggestions they are no longer needed, the syspath is automatically altered in `test/__init__.py`.

File: test/__init__.py (hunk 7 above)
Comment (m-horky) [reply test/__init__.py]: Done.

File: test/test_dnf_content_plugin.py (hunk 9 above)
Comment (m-horky) [reply test/test_dnf_content_plugin.py]: Ok.

**Comment by m-horky** [comment]
Thanks for the feedback. I've updated the PR with your suggested changes.

File: test-requirements.txt (hunk 13 above)
[the remark this answers is not in the document]
Comment (ptoscano) [reply test-requirements.txt:18]: hm fair enough

**Comment by m-horky** [comment]
The coverage should include all subdirectories of `src/` now.

**Review by ptoscano — APPROVED** [review]
LGTM, thanks! :+1:

**Review by jirihnidek — APPROVED** [review]
Thanks for updates. :+1: It works for me! It can go to main branch.
